server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java

```diff
@@ -426,6 +427,48 @@ public void testThresholdLimit() {
         );
     }
 
+    public void testShardSizeDiscrepancyWithinIndex() {
+        var discoveryNodesBuilder = DiscoveryNodes.builder();
+        for (int node = 0; node < 3; node++) {
+            discoveryNodesBuilder.add(newNode("node-" + node));
+        }
+
+        var metadataBuilder = Metadata.builder();
+        var routingTableBuilder = RoutingTable.builder();
+
+        addIndex(metadataBuilder, routingTableBuilder, "testindex", Map.of("node-0", 1, "node-1", 1));
+
+        var clusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .nodes(discoveryNodesBuilder)
+            .metadata(metadataBuilder)
+            .routingTable(routingTableBuilder)
+            .build();
+
+        var index = clusterState.routingTable().index("testindex").getIndex();
+
+        // Even if the two shards in this index vary massively in size, we must compute the balancing threshold to be high enough that
+        // we don't make pointless movements. 500GiB of difference is enough to demonstrate the bug.
+
+        var allocationService = createAllocationService(
```

Comment:
I noticed that this is creating allocation service with random ShardsAllocator. 
Should we fix it (and others `createAllocationService` in this file) to only use BalancedShardsAllocator?